Treat UAVS diffrently from manned aviation

LowOrbitIonCannon opened the pull request:
Using ADSB transmitter data existing or not.

A seperation distance of 10 Meters should suffice for D2D Collision avoidance. This is needed to test the collision avoidance of UAVs. 500 Meters is an unnecessary Distance for UAVs.

**Describe problem solved by the proposed pull request**
500 Meters is an unnecessary Distance for UAV to UAV Collision avoidance


**Describe your preferred solution**
A Flag that decides if the encountered entity is Manned or an UAV



**Additional context**
Collision avoidance uses UTM_Global_Position as per this [PR](https://github.com/PX4/Firmware/pull/12452)
Testing Collision avoidance would mean a seperation distance of 500 meters.


@hamishwillee  rightly informed me of the docs changes needed with this request:
http://docs.px4.io/master/en/advanced_features/traffic_avoidance_adsb.html#implementation

#11262

hamishwillee commented:
@ItsTimmy Don't know if you have interest in this since you looked at this code recently.

LorenzMeier commented:
Hey! Sorry, I've been traveling (and still are). What is your testing state with this PR?

LowOrbitIonCannon commented:
I wasnt yet able to test, I hope I can do that this week.

LowOrbitIonCannon commented:
I was able to Test today:

Tested was wether the Drone will Warn in QGC if there is another UAV nearby:

Testing Drone Traffic Warning with a Drone and a Rover

Test Setup:
Drone with D2X Module and QGC Groundstation. 
Rover with D2X Module and D2X Groundstation.
Modified Firmware running only on the Drone
"NAV_TRAFFIC_AVOID" in Warn Mode
MavLink Messages where send over Telem. 2 port to D2X.

Drone on NXP building terrace one and Rover on terrace two, distance between both is greater than 10 Meters.

Setup Pictures:
![20191029_124824](https://user-images.githubusercontent.com/5675438/67770271-57e0ea00-fa56-11e9-8012-fdd76c226966.jpg)
![Colli Test Setup Terrace 2 29 10 2019](https://user-images.githubusercontent.com/5675438/67772017-a479f480-fa59-11e9-9b55-f22cdc797c25.jpeg)

Stage 2:
![Colli Test Setup Stage 2 29 10 2019](https://user-images.githubusercontent.com/5675438/67771929-7e545480-fa59-11e9-995b-2a2f638f57f2.jpeg)

Test:
Stage 1:
Drone and Rover where switched on and started sending and receiving MavLink Messages.
Rover was emitting Collision warnings. Drone did not send those Messages:
<img width="769" alt="Colli Test No Warning 29 10 19 " src="https://user-images.githubusercontent.com/5675438/67769382-ac836580-fa54-11e9-83d1-281cd8581f7f.PNG">
![Colli Test Rover Groundstation Warning 29 10 2019](https://user-images.githubusercontent.com/5675438/67772030-aa6fd580-fa59-11e9-9f42-fce9f3cba8e3.jpeg)


Stage 2:
Rover moved over to terrace one, upon regain of GPS lock, the Drone was emitting Traffic Warnings:
<img width="777" alt="Colli Test Warning 29 10 19 " src="https://user-images.githubusercontent.com/5675438/67769535-ebb1b680-fa54-11e9-8eea-84cfd11d3205.PNG">
![Colli Test Rover Groundstation Warnings 29 10 2019](https://user-images.githubusercontent.com/5675438/67772038-ae035c80-fa59-11e9-907f-9b9918516599.jpeg)


Test was therefore successful.

dk7xe commented:
> Hey! Sorry, I've been traveling (and still are). What is your testing state with this PR?

Hi @LorenzMeier testing was successful. See https://github.com/PX4/Firmware/pull/13159#issuecomment-547424838
Who can take care on moving that forward?

ItsTimmy commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -960,6 +960,13 @@ void Navigator::check_traffic()
 
+
+		//ADSB Transmitter 14 = UAV
+		if (tr.emitter_type == 14) {
```
Can you use the `ADSB_EMITTER_TYPE_UAV` constant here?

ItsTimmy commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -960,6 +960,13 @@ void Navigator::check_traffic()
+		//ADSB Transmitter 14 = UAV
+		if (tr.emitter_type == 14) {
+			horizontal_separation = 10;
+			vertical_separation = 10;
```
Should these perhaps be parameters, instead of hard-coded? (This question is also valid for the separation distance for manned vehicles, which is currently hard-coded as 500 meters)

LowOrbitIonCannon commented:
I´ll test this with Timothy´s suggestions again on Wednesday.

LowOrbitIonCannon commented:
When i referenced a Mavlink Variable in Navigator_Main, Included Mavlink in the Cmake.txt. Now Px4 builds correctly but i get this Error when i Build nxp_fmuk66-v3:
![image](https://user-images.githubusercontent.com/5675438/68660292-2decf480-0539-11ea-981f-7a8e4eab73f2.png)

@davids5 could you take a look at this?

LowOrbitIonCannon commented:
I Added a simulated drone to fake_traffic and used that to validate my changes.
[Here ist the flight Log.](https://logs.px4.io/plot_app?log=e290b8cb-aeda-49cc-8d16-4de1ca2c0533)

What i tested: 

Standard Params
Start with Warn only.
NAV_TRAFF_AVOID =1
NAV_TRAFF_A_RADM= 500
NAV_TRAFF_A_RADM= 10

No warnings
Start with Warn only.
NAV_TRAFF_AVOID =1
NAV_TRAFF_A_RADM= 40
NAV_TRAFF_A_RADM= 5

All Warnings
Start with Warn only.
NAV_TRAFF_AVOID =1
NAV_TRAFF_A_RADM= 15000
NAV_TRAFF_A_RADM= 50


RETURN TO LAUNCH
NAV_TRAFF_AVOID =2
NAV_TRAFF_A_RADM= 15000
NAV_TRAFF_A_RADM= 50

LAND MODE
NAV_TRAFF_AVOID =3
NAV_TRAFF_A_RADM= 15000
NAV_TRAFF_A_RADM= 50


NO WARNING
NAV_TRAFF_AVOID =0
NAV_TRAFF_A_RADM= 15000
NAV_TRAFF_A_RADM= 50

All Messages where cut of at 50 Digits, in QGC, due to this limitation.
The log displays the Full message.
[MAVLINK STATUSTEXT](url)

NAV_TRAFFIC_AVOID = 3 is not implemented yet. So won't do anything.
(Land Mode)

![image](https://user-images.githubusercontent.com/5675438/69226353-49cb4880-0b80-11ea-9730-245f10c132c4.png)

This shows that the Parameter works correctly and Changes the distance PX4 Considers a Collision warning.

LowOrbitIonCannon commented on src/modules/navigator/CMakeLists.txt:
```diff
@@ -34,6 +34,10 @@
 	MODULE modules__navigator
 	MAIN navigator
+	COMPILE_FLAGS
+		-Wno-cast-align # TODO: fix and enable
```
This is how it is Implemented in the  CMakeLists.txt off mavlink_receiver.cpp.
This will need fixfin in the future

davids5 commented:
@mrpollo - Please add for dev call

dagar commented on src/modules/navigator/CMakeLists.txt:
```diff
@@ -34,6 +34,10 @@
+	COMPILE_FLAGS
+		-Wno-cast-align # TODO: fix and enable
+	INCLUDES
+		${PX4_SOURCE_DIR}/mavlink/include/mavlink
```
We don't want to directly access mavlink outside of the mavlink module.

LowOrbitIonCannon commented:
@dagar i added the variables to Transponder message

LowOrbitIonCannon commented:
I tested this again with the same Testsetup as above.
[LOG](https://review.px4.io/plot_app?log=ef76b019-d276-423a-9db1-36fb57e4c13a)

everything works as expected

LowOrbitIonCannon commented:
Added Behavior: Hold Position to collision avoidance mode.
and implemented Landmode to collision avoidance.

Landmode was not implemented in PX4.
Hold Position should be an available option to avoid Collisions


The German DLR wanted the Hold Position function for a Demo


While testing I found out that Check_Traffic will respond to selfpublished Uorb messages of Transponder_report.

hamishwillee commented:
FYI only, last set of docs discussed for this feature are here: https://github.com/PX4/px4_user_guide/pull/527. Still waiting on clarification.

hamishwillee commented:
I like the PR, but I expect a 10m minimum separation is too small. Drones can be moving quite fast, and there won't be time to react unless the transponder publish rate is very high.

LowOrbitIonCannon commented:
I had an interesting Testing day.

I changed where the UTM message is captured and ignored:
before it is send to Uorb to reduce Ressource usage.
These self published messages are reliably ignored now.

while Testing on the ground with 2 Vehicles. 
@dk7xe and me found out that the Parameter set in QGC for `nav_traff_radx` is off by a factor of 10.

We had trouble getting Collision messages on the modified Drone.
When we set down Drone 2 at a Distanz of 4 meters we could only get Collision messages: 
40<`nav_traff_radx` 
more testing Confirmed the issues we where having today.

Iam not sure why this happens. Is this due to the way the Parameter is Defined in QGC?


To warn the user I programmed the uas_id parameter into the mavlink_warning

[Flight log](https://logs.px4.io/plot_app?log=9bf7d812-9ea4-4fc0-a078-de709f6963fe)

LowOrbitIonCannon commented:
Boards with no HW `GUID `will send 0 as their `UAS_ID`.

This means that Collision avoidance between two drones that dont have a HW` GUID ` will not be processed.

This should be written down in Documentation.
And/Or maybe consider adding a Randomly generated HW `UID` as a Placeholder to Legacy Boards?
![image](https://user-images.githubusercontent.com/5675438/70611870-2c7d1d80-1c06-11ea-8e97-5db082528c3d.png)
This is the solution for no HW UID. maybe replace this.

LowOrbitIonCannon commented:
Succsessfully Tested.

[LOG](https://review.px4.io/plot_app?log=73c53122-c1a0-45f6-9901-f10faf8c4943)

Test was wether the Drone reacts correctly to changes in `Nav_traff_Rad`
and at what distanced.

Observed could be that the Parameter defined in QGC changes the maximum Distanz a Collision is called.
Self Published UTM_global_position messages where successfully ignored.

hamishwillee commented:
@julianoes FYI I am not tracking this discussion. Happy to rejoin when there are docs to be done.

LowOrbitIonCannon commented:
[LOG](https://logs.px4.io/plot_app?log=9ae5cb38-cff9-414b-a7f8-d0954fa4161c)
This log Does not show the Mavlink Error messages anymore.
They where caused by missmatched Submodules to the newer build.
Updating the subs removed the Errors.

Everything works now. @dagar could you Review it please?

I already started with the Updates to the Docs and will make a PR when this is approved.

dagar commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -1030,6 +1068,36 @@ void Navigator::check_traffic()
+							// ask the commander to land
+							vehicle_command_s vcmd = {};
+							vcmd.command = vehicle_command_s::VEHICLE_CMD_NAV_LAND;
+							publish_vehicle_cmd(&vcmd);
```
This is okay as a test, but if we want it to exist long term it should really be tied into the state machine a little more directly.

dagar commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -898,7 +898,7 @@ Navigator::load_fence_from_file(const char *filename)
  */
 void Navigator::fake_traffic(const char *callsign, float distance, float direction, float traffic_heading,
-			     float altitude_diff, float hor_velocity, float ver_velocity)
+			     float altitude_diff, float hor_velocity, float ver_velocity, int emitter_type)
```
```suggestion
			     float altitude_diff, float hor_velocity, float ver_velocity, uint16_t emitter_type)
```

Type should match the size used in the message definition.

dagar commented:
I didn't go through the specific utm_global_position handling in detail, but overall from PX4/Firmware perspective this change looks like safe to merge.

hamishwillee commented on src/modules/navigator/navigator_params.c:
```diff
@@ -121,11 +121,39 @@ PARAM_DEFINE_FLOAT(NAV_MC_ALT_RAD, 0.8f);
 PARAM_DEFINE_INT32(NAV_TRAFF_AVOID, 1);
 
+/**
+ * Set NAV TRAFFIC AVOID RADIUS MANNED
```
@LowOrbitIonCannon Can we have a new PR to change this to more natural formatting?

> Set avoidance radius/distance for manned traffic (ADS-B/FLARM).

hamishwillee commented on src/modules/navigator/navigator_params.c:
```diff
@@ -121,11 +121,39 @@ PARAM_DEFINE_FLOAT(NAV_MC_ALT_RAD, 0.8f);
+/**
+ * Set NAV TRAFFIC AVOID RADIUS MANNED
+ *
+ * Defines the Radius where NAV TRAFFIC AVOID is Called
```
@LowOrbitIonCannon Can we have a new PR to change this to more natural formatting?

> Defines the avoidance radius/distance at which the avoidance action is triggered for for manned vehicles (ADS-B/FLARM).

hamishwillee commented on src/modules/navigator/navigator_params.c:
```diff
@@ -121,11 +121,39 @@ PARAM_DEFINE_FLOAT(NAV_MC_ALT_RAD, 0.8f);
+PARAM_DEFINE_FLOAT(NAV_TRAFF_A_RADM, 500);
+
+/**
+ * Set NAV TRAFFIC AVOID RADIUS
```
@LowOrbitIonCannon Can we have a new PR to change this to more natural formatting?

... following pattern as I put above
